relay-server/src/utils/envelope_context.rs

```diff
@@ -41,10 +54,27 @@ impl EnvelopeContext {
             event_id: envelope.event_id(),
             remote_addr: meta.client_addr(),
             scoping: meta.get_partial_scoping(),
+            slot,
             done: false,
         }
     }
 
+    /// Creates a standalone `EnvelopeContext` for testing purposes.
+    ///
+    /// As opposed to [`new`](Self::new), this does not require a queue permit. This makes it
+    /// suitable for unit testing internals of the processing pipeline.
+    #[cfg(test)]
+    pub fn standalone(envelope: &Envelope) -> Self {
```

Comment:
I'm a bit mixed on this, you introduce some optional state - and importantly associated complexity - purely to avoid having to get a semaphore permit in a couple of tests.  I don't think getting a valid SemaphorePermit is that difficult to do in tests, so I'm mildly in favour of not introducing the complexity of the `Option` here.

Reply:
The optional state is needed for the drop internals already, so this is not added complexity. Also, having an option on internals is barely complexity, there's nothing actually complicated.

We can additionally get a proper permit in tests and introduce a utility for that, would that read better in tests?